File: java/src/test/java/com/penapereira/patterns/simplefactory/SimpleFactoryTest.java
package com.penapereira.patterns.simplefactory;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

class SimpleFactoryTest {

    private final SimpleFactory factory = new SimpleFactory();

    @Test
    void mapsTypeCodesToConcreteProducts() {
        assertInstanceOf(ConcreteProductA.class, factory.createProduct("A"));
        assertInstanceOf(ConcreteProductB.class, factory.createProduct("B"));
        assertEquals("ConcreteProductA", factory.createProduct("A").name());
        assertEquals("ConcreteProductB", factory.createProduct("B").name());
    }

    @Test
    void rejectsUnknownTypes() {
        IllegalArgumentException e =
                assertThrows(IllegalArgumentException.class, () -> factory.createProduct("Z"));
        assertEquals("Unknown product type: Z", e.getMessage());
    }
}

File: javascript/src/simple-factory/simple-factory.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { ConcreteProductA } from "./concrete-product-a.js";
import { ConcreteProductB } from "./concrete-product-b.js";
import { SimpleFactory } from "./simple-factory.js";

describe("SimpleFactory", () => {
  const factory = new SimpleFactory();

  it("maps type codes to concrete products", () => {
    assert.ok(factory.createProduct("A") instanceof ConcreteProductA);
    assert.ok(factory.createProduct("B") instanceof ConcreteProductB);
    assert.equal(factory.createProduct("A").name(), "ConcreteProductA");
    assert.equal(factory.createProduct("B").name(), "ConcreteProductB");
  });

  it("rejects unknown types", () => {
    assert.throws(() => factory.createProduct("Z"), /Unknown product type: Z/);
  });
});
